op-node/p2p/discovery.go

```diff
@@ -43,8 +82,295 @@ func (conf *Config) Discovery(log log.Logger) (*enode.LocalNode, *discover.UDPv5
+	chainID uint64
+	version uint64
+}
+
+func (o *OptimismENRData) ENRKey() string {
+	return "optimism"
+}
+
+func (o *OptimismENRData) EncodeRLP(w io.Writer) error {
+	out := make([]byte, 2*binary.MaxVarintLen64)
+	offset := binary.PutUvarint(out, o.chainID)
+	offset += binary.PutUvarint(out[offset:], o.version)
+	out = out[:offset]
+	// encode as byte-string
+	return rlp.Encode(w, out)
+}
+
+func (o *OptimismENRData) DecodeRLP(s *rlp.Stream) error {
+	b, err := s.Bytes()
+	if err != nil {
+		return fmt.Errorf("failed to decode outer ENR entry: %v", err)
+	}
+	// We don't check the byte length: the below readers are limited, and the ENR itself has size limits.
+	// Future "optimism" entries may contain additional data, and will be tagged with a newer version etc.
+	r := bytes.NewReader(b)
+	chainID, err := binary.ReadUvarint(r)
+	if err != nil {
+		return fmt.Errorf("failed to read chain ID var int: %v", err)
+	}
+	version, err := binary.ReadUvarint(r)
+	if err != nil {
+		return fmt.Errorf("failed to read version var int: %v", err)
+	}
+	o.chainID = chainID
+	o.version = version
+	return nil
+}
+
+var _ enr.Entry = (*OptimismENRData)(nil)
+
+func FilterEnodes(log log.Logger, cfg *rollup.Config) func(node *enode.Node) bool {
+	return func(node *enode.Node) bool {
+		var dat OptimismENRData
+		err := node.Load(&dat)
+		// if the entry does not exist, or if it is invalid, then ignore the node
+		if err != nil {
+			log.Debug("discovered node record has no optimism info", "node", node.ID(), "err", err)
+			return false
+		}
+		// check chain ID matches
+		if cfg.L2ChainID.Uint64() != dat.chainID {
+			log.Debug("discovered node record has no matching chain ID", "node", node.ID(), "got", dat.chainID, "expected", cfg.L2ChainID.Uint64())
+			return false
+		}
+		// check version matches
+		if dat.version != 0 {
+			log.Debug("discovered node record has no matching version", "node", node.ID(), "got", dat.version, "expected", 0)
+			return false
+		}
+		return true
+	}
+}
+
+// DiscoveryProcess runs a discovery process that randomly walks the DHT to fill the peerstore,
+// and connects to nodes in the peerstore that we are not already connected to.
+// Nodes from the peerstore will be shuffled, unsuccessful connection attempts will cause peers to be avoided,
+// and only nodes with addresses (under TTL) will be connected to.
+func (n *NodeP2P) DiscoveryProcess(ctx context.Context, log log.Logger, cfg *rollup.Config, connectGoal uint) {
+	if n.dv5Udp == nil {
+		log.Warn("peer discovery is disabled")
+		return
+	}
+	filter := FilterEnodes(log, cfg)
+	// We pull nodes from discv5 DHT in random order to find new peers.
+	// Eventually we'll find a peer record that matches our filter.
+	randomNodeIter := n.dv5Udp.RandomNodes()
+
+	randomNodeIter = enode.Filter(randomNodeIter, filter)
+	defer randomNodeIter.Close()
+
+	// We pull from the DHT in a slow/fast interval, depending on the need to find more peers
+	discoverTicker := time.NewTicker(time.Second * 5)
```

Comment:
Should be constants to start (IDK if we want to parameterize slow vs faster later one)